src/pk.c

```diff
@@ -6959,793 +7110,1199 @@ WOLFSSL_BIGNUM* wolfSSL_DH_8192_prime(WOLFSSL_BIGNUM* bn)
-               if (wolfSSL_BN_bin2bn(pub, pubSz, dh->pub_key) == NULL)
-                   WOLFSSL_MSG("Bad DH bn2bin error pub");
-               else if (wolfSSL_BN_bin2bn(priv, privSz, dh->priv_key) == NULL)
-                   WOLFSSL_MSG("Bad DH bn2bin error priv");
-               else
-                   ret = 1;
+            else {
+                /* Move pointer on if encoded. */
+                *out += len;
             }
         }
-        PRIVATE_KEY_LOCK();
     }
 
-    if (initTmpRng)
-        wc_FreeRng(tmpRng);
-
-#ifdef WOLFSSL_SMALL_STACK
-    XFREE(tmpRng, NULL, DYNAMIC_TYPE_RNG);
+    return (int)len;
 #endif
-    XFREE(pub,    NULL, DYNAMIC_TYPE_PUBLIC_KEY);
-    XFREE(priv,   NULL, DYNAMIC_TYPE_PRIVATE_KEY);
-
-    return ret;
 }
 
+#endif /* OPENSSL_ALL */
 
-/* return code compliant with OpenSSL :
- *   size of shared secret if success, -1 if error
+#endif /* !NO_CERTS */
+
+#endif /* OPENSSL_EXTRA */
+
+#if defined(OPENSSL_EXTRA) ||  \
+ ((!defined(NO_BIO) || !defined(NO_FILESYSTEM)) && \
+  defined(HAVE_LIGHTY) || defined(HAVE_STUNNEL) || \
+  defined(WOLFSSL_MYSQL_COMPATIBLE))
+
+/* Load the DER encoded DH parameters into DH key.
+ *
+ * @param [in, out] dh      DH key to load parameters into.
+ * @param [in]      derBuf  Buffer holding DER encoded parameters data.
+ * @param [in]      derSz   Size of DER data in buffer in bytes.
+ *
+ * @return  1 on success.
+ * @return  -1 when DH or derBuf is NULL,
+ *                  internal DH key in DH is NULL,
+ *                  derSz is 0 or less,
+ *                  error decoding DER data or
+ *                  setting external parameter values fails.
  */
-int wolfSSL_DH_compute_key(unsigned char* key, const WOLFSSL_BIGNUM* otherPub,
-                          WOLFSSL_DH* dh)
+int wolfSSL_DH_LoadDer(WOLFSSL_DH* dh, const unsigned char* derBuf, int derSz)
 {
-    int            ret    = -1;
-    word32         keySz  = 0;
-    int            pubSz  = 1024;
-    int            privSz = 1024;
-#ifdef WOLFSSL_SMALL_STACK
-    unsigned char* pub;
-    unsigned char* priv   = NULL;
+    int    ret = 1;
+#if !defined(HAVE_FIPS) || FIPS_VERSION_GT(2,0)
+    word32 idx = 0;
 #else
-    unsigned char  pub [1024];
-    unsigned char  priv[1024];
+    byte* p = NULL;
+    byte* g = NULL;
+    word32 pSz = MAX_DH_SIZE;
+    word32 gSz = MAX_DH_SIZE;
 #endif
 
-    WOLFSSL_MSG("wolfSSL_DH_compute_key");
+    /* Validate parameters. */
+    if ((dh == NULL) || (dh->internal == NULL) || (derBuf == NULL) ||
+            (derSz <= 0)) {
+        WOLFSSL_ERROR_MSG("Bad function arguments");
+        ret = -1;
+    }
 
-#ifdef WOLFSSL_SMALL_STACK
-    pub = (unsigned char*)XMALLOC(pubSz, NULL, DYNAMIC_TYPE_PUBLIC_KEY);
-    if (pub == NULL)
-        return ret;
+#if !defined(HAVE_FIPS) || FIPS_VERSION_GT(2,0)
+    /* Decode the parameters/key into the internal DH key. */
+    if ((ret == 1) && (wc_DhKeyDecode(derBuf, &idx, (DhKey*)dh->internal,
```

Comment:
This code is repeated at line 7167. Consider reducing duplication.

Reply:
Extracted.